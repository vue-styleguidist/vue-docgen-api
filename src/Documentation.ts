import Map from 'ts-map'

export type BlockTag = ParamTag | Tag

export interface ParamType {
  name: string
  elements?: ParamType[]
}

export interface UnnamedParam {
  type?: ParamType
  description?: string | boolean
}

export interface Param extends UnnamedParam {
  name?: string
}

interface RootTag {
  title: string
}

export interface Tag extends RootTag {
  content: string | boolean
}

export interface ParamTag extends RootTag, Param {}

export interface DocBlockTags {
  description: string
  tags?: Array<ParamTag | Tag>
}

interface EventType {
  names: string[]
}

interface EventProperty {
  type: EventType
  name?: string
  description?: string | boolean
}

export interface EventDescriptor extends DocBlockTags {
  type?: EventType
  properties: EventProperty[] | undefined
}

export interface PropDescriptor {
  type?: { name: string; func?: boolean }
  description: string
  required?: string | boolean
  defaultValue?: { value: string; func?: boolean }
  tags: { [title: string]: BlockTag[] }
}

export interface MethodDescriptor {
  name: string
  description: string
  returns?: UnnamedParam
  tags?: { [key: string]: BlockTag[] }
  params?: Param[]
  modifiers: string[]
  [key: string]: any
}

export interface SlotDescriptor {
  description?: string
  bindings?: Record<string, any>
  scoped?: boolean
}

export interface ComponentDoc {
  displayName: string
  props: { [propName: string]: PropDescriptor } | undefined
  methods: MethodDescriptor[]
  slots: { [name: string]: SlotDescriptor }
  events?: { [name: string]: EventDescriptor }
  tags: { [key: string]: BlockTag[] }
  [key: string]: any
}

export class Documentation {
  private propsMap: Map<string, PropDescriptor>
  private methodsMap: Map<string, MethodDescriptor>
  private slotsMap: Map<string, SlotDescriptor>
  private eventsMap: Map<string, any>
  private dataMap: Map<string, any>
  constructor() {
    this.propsMap = new Map()
    this.methodsMap = new Map()
    this.slotsMap = new Map()
    this.eventsMap = new Map()

    this.dataMap = new Map()
  }

  public set(key: string, value: any) {
    this.dataMap.set(key, value)
  }

  public get(key: string): any {
    return this.dataMap.get(key)
  }

  public getPropDescriptor(propName: string): PropDescriptor {
    return this.getDescriptor(propName, this.propsMap, () => ({
      description: '',
      tags: {},
    }))
  }

  public getMethodDescriptor(methodName: string): MethodDescriptor {
    return this.getDescriptor(methodName, this.methodsMap, () => ({
      name: methodName,
      modifiers: [],
      description: '',
      tags: {},
    }))
  }

  public getEventDescriptor(eventName: string): EventDescriptor {
    return this.getDescriptor(eventName, this.eventsMap, () => ({
      properties: undefined,
      description: '',
      tags: [],
    }))
  }

  public getSlotDescriptor(slotName: string): SlotDescriptor {
    return this.getDescriptor(slotName, this.slotsMap, () => ({
      description: '',
    }))
  }

  public toObject(): ComponentDoc {
    const props = this.getPropsObject()
    const methods = this.getMethodsObject()
    const events = this.getEventsObject() || {}
    const slots = this.getSlotsObject()

    const obj: { [key: string]: any } = {}
    this.dataMap.forEach((value, key) => {
      if (key) {
        obj[key] = value
      }
    })

    return {
      ...obj,
      // initialize non null params
      description: obj.description || '',
      tags: obj.tags || {},

      // set all the static properties
      displayName: obj.displayName,
      props,
      events,
      methods,
      slots,
    }
  }

  private getDescriptor<T>(name: string, map: Map<string, T>, init: () => T): T {
    let descriptor = map.get(name)
    if (!descriptor) {
      map.set(name, (descriptor = init()))
    }
    return descriptor
  }

  private getObjectFromDescriptor<T>(map: Map<string, T>): { [name: string]: T } | undefined {
    if (map.size > 0) {
      const obj: { [name: string]: T } = {}
      map.forEach((descriptor, name) => {
        if (name && descriptor) {
          obj[name] = descriptor
        }
      })
      return obj
    } else {
      return undefined
    }
  }

  private getPropsObject(): { [propName: string]: PropDescriptor } | undefined {
    return this.getObjectFromDescriptor(this.propsMap)
  }

  private getMethodsObject(): MethodDescriptor[] {
    const methods: MethodDescriptor[] = []
    this.methodsMap.forEach((descriptor, name) => {
      if (name && methods && descriptor) {
        methods.push(descriptor)
      }
    })
    return methods
  }

  private getEventsObject(): { [eventName: string]: EventDescriptor } | undefined {
    return this.getObjectFromDescriptor(this.eventsMap)
  }

  private getSlotsObject(): { [slotName: string]: SlotDescriptor } {
    return this.getObjectFromDescriptor(this.slotsMap) || {}
  }
}
