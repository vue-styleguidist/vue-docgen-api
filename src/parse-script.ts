import { ParserPlugin } from '@babel/parser'
import * as bt from '@babel/types'
import { NodePath } from 'ast-types'
import Map from 'ts-map'
import buildParser from './babel-parser'
import { Documentation } from './Documentation'
import cacher from './utils/cacher'
import resolveExportedComponent from './utils/resolveExportedComponent'

// tslint:disable-next-line:no-var-requires
import recast = require('recast')

const ERROR_MISSING_DEFINITION = 'No suitable component definition found'

interface ParseScriptOptions {
  lang: 'ts' | 'js'
  filePath: string
  nameFilter?: string[]
}

export default function parseScript(
  source: string,
  documentation: Documentation,
  handlers: Array<
    (doc: Documentation, componentDefinition: NodePath, ast: bt.File, filePath: string) => void
  >,
  options: ParseScriptOptions,
) {
  const plugins: ParserPlugin[] = options.lang === 'ts' ? ['typescript'] : ['flow']

  const ast = cacher(() => recast.parse(source, { parser: buildParser({ plugins }) }), source)
  if (!ast) {
    throw new Error(ERROR_MISSING_DEFINITION)
  }

  const componentDefinitions = resolveExportedComponent(ast)

  if (componentDefinitions.size === 0) {
    throw new Error(ERROR_MISSING_DEFINITION)
  }

  executeHandlers(handlers, componentDefinitions, documentation, ast, options)
}

function executeHandlers(
  localHandlers: Array<
    (doc: Documentation, componentDefinition: NodePath, ast: bt.File, filePath: string) => void
  >,
  componentDefinitions: Map<string, NodePath>,
  documentation: Documentation,
  ast: bt.File,
  opt: ParseScriptOptions,
) {
  return componentDefinitions.forEach((compDef, name) => {
    if (compDef && name && (!opt.nameFilter || opt.nameFilter.indexOf(name) > -1)) {
      localHandlers.forEach(handler => handler(documentation, compDef, ast, opt.filePath))
    }
  })
}
