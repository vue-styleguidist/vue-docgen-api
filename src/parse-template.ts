import * as pug from 'pug'
import { ASTElement, ASTNode, compile, SFCBlock } from 'vue-template-compiler'
import { Documentation } from './Documentation'
import cacher from './utils/cacher'

export interface TemplateParserOptions {
  functional: boolean
  rootLeadingComment: string
}

export type Handler = (
  documentation: Documentation,
  templateAst: ASTElement,
  options: TemplateParserOptions,
) => void

export default function parseTemplate(
  tpl: SFCBlock,
  documentation: Documentation,
  handlers: Handler[],
  filePath: string,
) {
  if (tpl && tpl.content) {
    const template =
      tpl.attrs && tpl.attrs.lang === 'pug'
        ? pug.render(tpl.content, { filename: filePath })
        : tpl.content
    const ast = cacher(() => compile(template, { comments: true }).ast, template)
    const rootLeadingCommentArray = /^<!--(.+)-->/.exec(template.trim())
    const rootLeadingComment =
      rootLeadingCommentArray && rootLeadingCommentArray.length > 1
        ? rootLeadingCommentArray[1].trim()
        : ''
    if (ast) {
      traverse(ast, documentation, handlers, {
        functional: !!tpl.attrs.functional,
        rootLeadingComment,
      })
    }
  }
}

export function traverse(
  templateAst: ASTElement,
  documentation: Documentation,
  handlers: Handler[],
  options: TemplateParserOptions,
) {
  if (templateAst.type === 1) {
    handlers.forEach(handler => {
      handler(documentation, templateAst, options)
    })
    if (templateAst.children) {
      for (const childNode of templateAst.children) {
        if (isASTElement(childNode)) {
          traverse(childNode, documentation, handlers, options)
        }
      }
    }
  }
}

function isASTElement(node: ASTNode): node is ASTElement {
  return !!node && (node as ASTElement).children !== undefined
}
