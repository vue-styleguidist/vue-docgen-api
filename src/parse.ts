import * as fs from 'fs'
import * as path from 'path'
import { parseComponent, SFCDescriptor } from 'vue-template-compiler'
import { Documentation } from './Documentation'
import parseScript from './parse-script'
import parseTemplate from './parse-template'
import handlers from './script-handlers'
import templateHandlers from './template-handlers'
import cacher from './utils/cacher'

const ERROR_EMPTY_DOCUMENT = 'The passed source is empty'

/**
 * parses the source and returns the doc
 * @param {string} source code whose documentation is parsed
 * @param {string} filePath path of the current file against whom to resolve the mixins
 * @returns {object} documentation object
 */
export function parseFile(filePath: string, documentation: Documentation) {
  const source = fs.readFileSync(filePath, {
    encoding: 'utf-8',
  })
  return parseSource(source, filePath, documentation)
}

/**
 * parses the source and returns the doc
 * @param {string} source code whose documentation is parsed
 * @param {string} filePath path of the current file against whom to resolve the mixins
 * @returns {object} documentation object
 */
export function parseSource(source: string, filePath: string, documentation: Documentation) {
  const singleFileComponent = /\.vue$/i.test(path.extname(filePath))
  let parts: SFCDescriptor | null = null

  if (source === '') {
    throw new Error(ERROR_EMPTY_DOCUMENT)
  }

  if (singleFileComponent) {
    parts = cacher(() => parseComponent(source), source)
  }

  const scriptSource = parts ? (parts.script ? parts.script.content : undefined) : source
  if (scriptSource) {
    const lang =
      (parts && parts.script && parts.script.attrs && parts.script.attrs.lang === 'ts') ||
      /\.tsx?$/i.test(path.extname(filePath))
        ? 'ts'
        : 'js'
    parseScript(scriptSource, documentation, handlers, { lang, filePath })
  }

  // get slots from template
  if (parts && parts.template) {
    parseTemplate(parts.template, documentation, templateHandlers, filePath)
  }

  if (!documentation.get('displayName')) {
    // a component should always have a display name
    documentation.set('displayName', path.basename(filePath).replace(/\.\w+$/, ''))
  }
}
