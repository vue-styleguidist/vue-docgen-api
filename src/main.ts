import { ComponentDoc, Documentation } from './Documentation'
import { parseFile, ParseOptions, parseSource as parseSourceLocal } from './parse'

export { ComponentDoc, ParseOptions }

export function parse(
  filePath: string,
  opts?: ParseOptions | { [alias: string]: string },
): ComponentDoc {
  const doc = new Documentation()
  const options: ParseOptions = isParserOptions(opts) ? opts : { filePath, aliases: opts }
  parseFile(doc, options)
  return doc.toObject()
}

export function parseSource(
  source: string,
  filePath: string,
  opts?: ParseOptions | { [alias: string]: string },
): ComponentDoc {
  const doc = new Documentation()
  const options: ParseOptions = isParserOptions(opts) ? opts : { filePath, aliases: opts }
  parseSourceLocal(doc, source, options)
  return doc.toObject()
}

function isParserOptions(opts: any): opts is ParseOptions {
  return !!opts && !!opts.aliases
}
