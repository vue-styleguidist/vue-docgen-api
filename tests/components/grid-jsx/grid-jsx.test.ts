import * as path from 'path'
import { ComponentDoc, PropDescriptor } from '../../../src/Documentation'
import { parse } from '../../../src/main'

const grid = path.join(__dirname, './Grid.vue')
let docGrid: ComponentDoc

describe('tests grid jsx', () => {
  beforeAll(done => {
    docGrid = parse(grid)
    done()
  })

  it('should return an object', () => {
    expect(typeof docGrid).toBe('null')
  })

  it('The component name should be grid', () => {
    expect(docGrid.displayName).toEqual('grid')
  })

  it('should the component has tags', () => {
    expect(typeof docGrid.tags !== 'undefined').toBe(true)
  })

  it('should the component has authors', () => {
    expect(typeof docGrid.tags.author !== 'undefined').toBe(true)
  })

  it('should the component has description', () => {
    expect(typeof docGrid.description !== 'undefined').toBe(true)
  })

  it('should has methods', () => {
    expect(typeof docGrid.methods !== 'undefined').toBe(true)
  })

  it('should the component has one method', () => {
    expect(Object.keys(docGrid.methods).length).toEqual(1)
  })

  it('should has props', () => {
    expect(docGrid.props).not.toBeUndefined()
  })

  it('should the component has version', () => {
    expect(typeof docGrid.tags.version !== 'undefined').toBe(true)
  })
  describe('props', () => {
    let props: { [propName: string]: PropDescriptor }

    beforeAll(() => {
      props = docGrid.props ? docGrid.props : {}
    })
    it('should the component has four props', () => {
      expect(Object.keys(props).length).toEqual(6)
    })

    it('grid component should have a msg prop as string|number type', () => {
      expect(props.msg.type).toMatchObject({ name: 'string|number' })
    })

    it('grid component should have a filterKey prop as string type', () => {
      expect(props.filterKey.type).toMatchObject({ name: 'string' })
    })

    it('grid component should have a propFunc prop as func type', () => {
      expect(props.propFunc.type).toMatchObject({ name: 'func' })
    })

    it('grid component should have a images prop as Array type', () => {
      expect(props.images.type).toMatchObject({ name: 'array' })
    })

    it('grid component should have a data prop as Array type', () => {
      expect(props.data.type).toMatchObject({ name: 'array' })
    })

    it('grid component should have a columns prop as Array type', () => {
      expect(props.columns.type).toMatchObject({ name: 'array' })
    })

    it('should the prop msg has four tags', () => {
      expect(Object.keys(props.msg.tags).length).toEqual(4)
    })
  })

  it('should the component has two event', () => {
    expect(Object.keys(docGrid.events || {}).length).toEqual(2)
  })

  it('should the component has event, it called success', () => {
    expect(typeof (docGrid.events || {}).success !== 'undefined').toBe(true)
  })

  it('should the description of success event is Success event.', () => {
    expect((docGrid.events || {}).success.description).toEqual('Success event.')
  })

  it('should the component has event, it called error', () => {
    expect(typeof (docGrid.events || {}).error !== 'undefined').toBe(true)
  })

  it('should the description of error event is Error event.', () => {
    expect((docGrid.events || {}).error.description).toEqual('Error event.')
  })

  it('should define the return type of the first method', () => {
    expect(docGrid.methods[0].returns).toMatchObject({ description: 'Test' })
  })

  it('should match the snapshot', () => {
    expect(docGrid).toMatchSnapshot()
  })
})
