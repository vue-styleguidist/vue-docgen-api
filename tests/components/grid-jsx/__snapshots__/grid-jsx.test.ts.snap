// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests grid jsx should match the snapshot 1`] = `
Object {
  "description": "This is an example of creating a reusable grid component and using it with external data.",
  "displayName": "grid",
  "events": Object {
    "error": Object {
      "description": "Error event.",
      "properties": undefined,
      "type": Object {
        "names": Array [
          "object",
        ],
      },
    },
    "success": Object {
      "description": "Success event.",
      "properties": undefined,
      "type": Object {
        "names": Array [
          "object",
        ],
      },
    },
  },
  "methods": Array [
    Object {
      "description": "Sets the order",
      "modifiers": Array [],
      "name": "sortBy",
      "params": Array [
        Object {
          "description": "Key to order",
          "name": "key",
          "type": Object {
            "name": "string",
          },
        },
      ],
      "returns": Object {
        "description": "Test",
        "title": "returns",
        "type": Object {
          "name": "string",
        },
      },
      "tags": Object {
        "access": Array [
          Object {
            "description": "public",
            "title": "access",
          },
        ],
        "params": Array [
          Object {
            "description": "Key to order",
            "name": "key",
            "title": "param",
            "type": Object {
              "name": "string",
            },
          },
        ],
        "returns": Array [
          Object {
            "description": "Test",
            "title": "returns",
            "type": Object {
              "name": "string",
            },
          },
        ],
        "since": Array [
          Object {
            "description": "Version 1.0.1",
            "title": "since",
          },
        ],
        "version": Array [
          Object {
            "description": "1.0.5",
            "title": "version",
          },
        ],
      },
    },
  ],
  "props": Object {
    "columns": Object {
      "description": "get columns list",
      "tags": Object {},
      "type": Object {
        "name": "array",
      },
    },
    "data": Object {
      "description": "describe data",
      "tags": Object {
        "version": Array [
          Object {
            "description": "1.0.5",
            "title": "version",
          },
        ],
      },
      "type": Object {
        "name": "array",
      },
    },
    "filterKey": Object {
      "defaultValue": Object {
        "func": false,
        "value": "'example'",
      },
      "description": "filter key",
      "required": "",
      "tags": Object {
        "ignore": Array [
          Object {
            "description": true,
            "title": "ignore",
          },
        ],
      },
      "type": Object {
        "name": "string",
      },
    },
    "images": Object {
      "defaultValue": Object {
        "func": true,
        "value": "function() {
  return [{}]
}",
      },
      "description": "",
      "required": "",
      "tags": Object {},
      "type": Object {
        "name": "array",
      },
    },
    "msg": Object {
      "defaultValue": Object {
        "func": false,
        "value": "text",
      },
      "description": "object/array defaults should be returned from a factory function",
      "required": "",
      "tags": Object {
        "link": Array [
          Object {
            "description": "See [Wikipedia](https://en.wikipedia.org/wiki/Web_colors#HTML_color_names) for a list of color names",
            "title": "link",
          },
        ],
        "see": Array [
          Object {
            "description": "See [Wikipedia](https://en.wikipedia.org/wiki/Web_colors#HTML_color_names) for a list of color names",
            "title": "see",
          },
        ],
        "since": Array [
          Object {
            "description": "Version 1.0.1",
            "title": "since",
          },
        ],
        "version": Array [
          Object {
            "description": "1.0.5",
            "title": "version",
          },
        ],
      },
      "type": Object {
        "name": "string|number",
      },
    },
    "propFunc": Object {
      "defaultValue": Object {
        "func": true,
        "value": "function() {}",
      },
      "description": "prop function",
      "required": "",
      "tags": Object {},
      "type": Object {
        "name": "func",
      },
    },
  },
  "slots": Object {
    "header": Object {
      "description": "Use this slot header",
    },
  },
  "tags": Object {
    "author": Array [
      Object {
        "description": "[Rafael](https://github.com/rafaesc92)",
        "title": "author",
      },
    ],
    "since": Array [
      Object {
        "description": "Version 1.0.1",
        "title": "since",
      },
    ],
    "version": Array [
      Object {
        "description": "1.0.5",
        "title": "version",
      },
    ],
  },
}
`;
